File: go_lib/bassa.go
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License

package bassa

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"regexp"
	"time"

	"github.com/gojektech/heimdall"
	"github.com/gojektech/heimdall/httpclient"
	"github.com/hokaccha/go-prettyjson"
)

// Bassa : Bassa Go object
type Bassa struct {
	apiURL     string
	token      string
	timeout    int
	retryCount int
	httpClient *httpclient.Client
}

var (
	errBadFormat        = errors.New("invalid format")
	errIncompleteParams = errors.New("Some fields are not valid or empty")
	emailRegexp         = regexp.MustCompile("^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$")
)

// validateFormat : Helper function to validate email address
func validateFormat(email string) error {
	if !emailRegexp.MatchString(email) {
		return errBadFormat
	}
	return nil
}

// Init : Initialization of Bassa
func (b *Bassa) Init(apiURL string, timeout int, retryCount int) {
	if apiURL == "" || timeout == 0 {
		panic(errIncompleteParams)
	}
	u, err := url.Parse(apiURL)
	if err != nil {
		fmt.Println(u)
		panic(err)
	} else {
		b.apiURL = apiURL
		b.timeout = timeout
		b.retryCount = retryCount
		b.token = ""
		timeout := time.Duration(timeout) * time.Millisecond
		httpClient := httpclient.NewClient(
			httpclient.WithHTTPTimeout(timeout),
			httpclient.WithRetryCount(retryCount),
			httpclient.WithRetrier(heimdall.NewRetrier(heimdall.NewConstantBackoff(10*time.Millisecond, 50*time.Millisecond))),
		)
		b.httpClient = httpClient
	}
}

// Login : Function to login as a user
func (b *Bassa) Login(userName string, password string) {
	if userName == "" || password == "" {
		panic(errIncompleteParams)
	}
	endpoint := "/api/login"
	apiURL := b.apiURL + endpoint

	form := url.Values{}
	form.Add("user_name", userName)
	form.Add("password", password)

	response, err := http.PostForm(apiURL, form)
	if err != nil {
		panic(err)
	}
	defer response.Body.Close()
	b.token = response.Header["Token"][0]

	respBody, err := ioutil.ReadAll(response.Body)
	if err != nil {
		fmt.Println(string(respBody))
		panic(err)
	}
}

// AddRegularUserRequest : Function to login as a user
func (b *Bassa) AddRegularUserRequest(userName string, password string, email string) {
	if userName == "" || password == "" || email == "" {
		panic(errIncompleteParams)
	}

	err := validateFormat(email)
	if err != nil {
		panic(err)
	}

	endpoint := "/api/regularuser"
	apiURL := b.apiURL + endpoint

	requestBody, err := json.Marshal(map[string]string{
		"user_name": userName,
		"password":  password,
		"email":     email})
	if err != nil {
		panic(err)
	}
	request, err := http.NewRequest("POST", apiURL, bytes.NewBuffer(requestBody))
	if err != nil {
		panic(err)
	}
	request.Header.Set("token", b.token)
	response, err := b.httpClient.Do(request)
	if err != nil {
		panic(err)
	}

	defer response.Body.Close()

	respBody, err := ioutil.ReadAll(response.Body)
	if err != nil {
		fmt.Println(string(respBody))
		panic(err)
	}
}

// AddUserRequest : Function to login as a user
func (b *Bassa) AddUserRequest(userName string, password string, email string, authLevel int) {
	if userName == "" || password == "" || email == "" {
		panic(errIncompleteParams)
	}

	err := validateFormat(email)
	if err != nil {
		panic(err)
	}

	endpoint := "/api/user"
	apiURL := b.apiURL + endpoint

	requestBody := []byte(fmt.Sprintf("{user_name:\"%s\", password: \"%s\", email: \"%s\", auth: %d}", userName, password, email, authLevel))

	request, err := http.NewRequest("POST", apiURL, bytes.NewBuffer(requestBody))
	if err != nil {
		panic(err)
	}
	request.Header.Set("token", b.token)
	response, err := b.httpClient.Do(request)
	if err != nil {
		panic(err)
	}

	defer response.Body.Close()
	var r interface{}
	if err := json.NewDecoder(response.Body).Decode(&r); err != nil {
		panic(err)
	}
	out, err := prettyjson.Marshal(r)
	fmt.Println(string(out))
}

// RemoveUserRequest : Function to remove user
func (b *Bassa) RemoveUserRequest(userName string) string {
	if userName == "" {
		panic(errIncompleteParams)
	}

	endpoint := "/api/user" + "/" + userName
	apiURL := b.apiURL + endpoint

	request, err := http.NewRequest("DELETE", apiURL, nil)
	if err != nil {
		panic(err)
	}
	request.Header.Set("token", b.token)
	response, err := b.httpClient.Do(request)
	if err != nil {
		panic(err)
	}

	defer response.Body.Close()
	var r interface{}
	if err := json.NewDecoder(response.Body).Decode(&r); err != nil {
		panic(err)
	}
	out, err := prettyjson.Marshal(r)
	return string(out)
}

// UpdateUserRequest : Function to update user request
func (b *Bassa) UpdateUserRequest(userName string, newUserName string, password string, authLevel int, email string) {
	if userName == "" || password == "" || email == "" || newUserName == "" {
		panic(errIncompleteParams)
	}

	err := validateFormat(email)
	if err != nil {
		panic(err)
	}

	endpoint := "/api/user"
	apiURL := b.apiURL + endpoint + "/" + userName

	requestBody := []byte(fmt.Sprintf("{user_name:\"%s\", password: \"%s\", email: \"%s\", auth_level: %d}", newUserName, password, email, authLevel))

	request, err := http.NewRequest("PUT", apiURL, bytes.NewBuffer(requestBody))
	if err != nil {
		panic(err)
	}
	request.Header.Set("token", b.token)
	response, err := b.httpClient.Do(request)
	if err != nil {
		panic(err)
	}

	defer response.Body.Close()
	var r interface{}
	if err := json.NewDecoder(response.Body).Decode(&r); err != nil {
		panic(err)
	}
	out, err := prettyjson.Marshal(r)
	fmt.Println(string(out))
}

// GetUserRequest : Function to get user request
func (b *Bassa) GetUserRequest() string {

	endpoint := "/api/user"
	apiURL := b.apiURL + endpoint

	request, err := http.NewRequest("GET", apiURL, nil)
	if err != nil {
		panic(err)
	}
	request.Header.Set("token", b.token)
	response, err := b.httpClient.Do(request)
	if err != nil {
		panic(err)
	}

	defer response.Body.Close()
	var r interface{}
	if err := json.NewDecoder(response.Body).Decode(&r); err != nil {
		panic(err)
	}
	out, err := prettyjson.Marshal(r)
	fmt.Println(string(out))
	return string(out)
}

// GetUserSignupRequests : Function to get user signup requests
func (b *Bassa) GetUserSignupRequests() string {

	endpoint := "/api/user/requests"
	apiURL := b.apiURL + endpoint

	request, err := http.NewRequest("GET", apiURL, nil)
	if err != nil {
		panic(err)
	}
	request.Header.Set("token", b.token)
	response, err := b.httpClient.Do(request)
	if err != nil {
		panic(err)
	}

	defer response.Body.Close()
	var r interface{}
	if err := json.NewDecoder(response.Body).Decode(&r); err != nil {
		panic(err)
	}
	out, err := prettyjson.Marshal(r)
	fmt.Println(string(out))
	return string(out)
}

// ApproveUserRequest : Function to approve user request
func (b *Bassa) ApproveUserRequest(userName string) {
	if userName == "" {
		panic(errIncompleteParams)
	}
	endpoint := "/api/user/approve"
	apiURL := b.apiURL + endpoint + "/" + userName

	request, err := http.NewRequest("POST", apiURL, nil)
	if err != nil {
		panic(err)
	}
	request.Header.Set("token", b.token)
	response, err := b.httpClient.Do(request)
	if err != nil {
		panic(err)
	}

	defer response.Body.Close()
	var r interface{}
	if err := json.NewDecoder(response.Body).Decode(&r); err != nil {
		panic(err)
	}
	out, err := prettyjson.Marshal(r)
	fmt.Println(string(out))
}

// GetBlockedUserRequests : Function to get blocked user requests
func (b *Bassa) GetBlockedUserRequests() string {

	endpoint := "/api/user/blocked"
	apiURL := b.apiURL + endpoint

	request, err := http.NewRequest("GET", apiURL, nil)
	if err != nil {
		panic(err)
	}
	request.Header.Set("token", b.token)
	response, err := b.httpClient.Do(request)
	if err != nil {
		panic(err)
	}

	defer response.Body.Close()
	var r interface{}
	if err := json.NewDecoder(response.Body).Decode(&r); err != nil {
		panic(err)
	}
	out, err := prettyjson.Marshal(r)
	fmt.Println(string(out))
	return string(out)
}

// BlockUserRequest : Function to block user request
func (b *Bassa) BlockUserRequest(userName string) {
	if userName == "" {
		panic(errIncompleteParams)
	}
	endpoint := "/api/user/blocked"
	apiURL := b.apiURL + endpoint + "/" + userName

	request, err := http.NewRequest("POST", apiURL, nil)
	if err != nil {
		panic(err)
	}
	request.Header.Set("token", b.token)
	response, err := b.httpClient.Do(request)
	if err != nil {
		panic(err)
	}

	defer response.Body.Close()
	var r interface{}
	if err := json.NewDecoder(response.Body).Decode(&r); err != nil {
		panic(err)
	}
	out, err := prettyjson.Marshal(r)
	fmt.Println(string(out))
}

// UnBlockUserRequest : Function to unblock user request
func (b *Bassa) UnBlockUserRequest(userName string) {
	if userName == "" {
		panic(errIncompleteParams)
	}
	endpoint := "/api/user/blocked"
	apiURL := b.apiURL + endpoint + "/" + userName

	request, err := http.NewRequest("DELETE", apiURL, nil)
	if err != nil {
		panic(err)
	}
	request.Header.Set("token", b.token)
	response, err := b.httpClient.Do(request)
	if err != nil {
		panic(err)
	}

	defer response.Body.Close()
	var r interface{}
	if err := json.NewDecoder(response.Body).Decode(&r); err != nil {
		panic(err)
	}
	out, err := prettyjson.Marshal(r)
	fmt.Println(string(out))
}

// GetDownloadUserRequests : Function to get download user requests
func (b *Bassa) GetDownloadUserRequests(limit int) string {
	if limit == 0 {
		limit = 1
	}
	endpoint := "/api/user/downloads"
	apiURL := b.apiURL + endpoint + "/" + string(limit)

	request, err := http.NewRequest("GET", apiURL, nil)
	if err != nil {
		panic(err)
	}
	request.Header.Set("token", b.token)
	response, err := b.httpClient.Do(request)
	if err != nil {
		panic(err)
	}

	defer response.Body.Close()
	var r interface{}
	if err := json.NewDecoder(response.Body).Decode(&r); err != nil {
		panic(err)
	}
	out, err := prettyjson.Marshal(r)
	fmt.Println(string(out))
	return string(out)
}

// GetToptenHeaviestUsers : Function to get top ten heaviest users
func (b *Bassa) GetToptenHeaviestUsers() string {

	endpoint := "/api/user/heavy"
	apiURL := b.apiURL + endpoint

	request, err := http.NewRequest("GET", apiURL, nil)
	if err != nil {
		panic(err)
	}
	request.Header.Set("token", b.token)
	response, err := b.httpClient.Do(request)
	if err != nil {
		panic(err)
	}

	defer response.Body.Close()
	var r interface{}
	if err := json.NewDecoder(response.Body).Decode(&r); err != nil {
		panic(err)
	}
	out, err := prettyjson.Marshal(r)
	fmt.Println(string(out))
	return string(out)
}

File: js_lib/bassa/bassa.js
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//       http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

// Bassa js client library will enable you to interact easily with Bassa API server with most of its functions covered. 

const axios = require('axios');
const { URL } = require('url');
const { InvalidUrl, IncompleteParams, ResponseError } = require('./errors');

class Bassa {
    constructor(apiUrl, total = 1, backoffFactor = 1, timeout = 5) {
        this.apiUrl = apiUrl;
        this.headers = {
            'Content-Type': 'application/x-www-form-urlencoded',
        };

        const reg = new RegExp(
            '^(?:http|ftp)s?://' +  // http:// or https://
            // domain...
            '(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\\.)+(?:[A-Z]{2,6}\\.?|[A-Z0-9-]{2,}\\.?)|' +
            'localhost|' +  // localhost...
            '\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3})' +  // ...or ip
            '(?::\\d+)?' +  // optional port
            '(?:/?|[/?]\\S+)$', 'i'
        );

        if (!reg.test(apiUrl)) {
            throw new InvalidUrl('Invalid URL');
        }

        const retryStrategy = {
            total,
            backoffFactor,
            retry: [429, 500, 502, 503, 504],
        };

        const instance = axios.create({
            baseURL: apiUrl,
            timeout: timeout * 1000,
            httpAgent: new axios.Agent({
                http: new axios.Agent(retryStrategy),
                https: new axios.Agent(retryStrategy),
            }),
        });

        this.http = instance;
    }

    // User Functions

    login(user_name, password) {
        const endpoint = "/api/login";
        const url = new URL(endpoint, this.apiUrl);
        const params = new URLSearchParams();
        if (!user_name || !password) {
            throw new IncompleteParams();
        }
        params.append('user_name', user_name);
        params.append('password', password);

        return this.http.post(url.href, params, { headers: this.headers })
            .then(response => {
                if (response.status === 200) {
                    this.headers['token'] = response.headers.token;
                } else {
                    throw new ResponseError(`API response: ${response.status}`);
                }
            });
    }

    addRegularUserRequest(user_name, password, email) {
        const endpoint = "/api/regularuser";
        const url = new URL(endpoint, this.apiUrl);
        const params = new URLSearchParams();
        if (!user_name || !password || !email) {
            throw new IncompleteParams();
        }
        params.append('user_name', user_name);
        params.append('password', password);
        params.append('email', email);
    
        return this.http.post(url.href, params, { headers: this.headers });
    }

    addUserRequest(user_name, password, email, auth_level = 1) {
        const endpoint = "/api/user";
        const url = new URL(endpoint, this.apiUrl);
        const params = new URLSearchParams();
        if (!user_name || !password || !email) {
            throw new IncompleteParams();
        }
        params.append('user_name', user_name);
        params.append('password', password);
        params.append('email', email);
        params.append('auth', auth_level);
    
        return this.http.post(url.href, params, { headers: this.headers });
    }

    removeUserRequest(user_name) {
        if (!user_name) {
            throw new IncompleteParams();
        }
        const endpoint = "/api/user";
        const url = new URL(`${endpoint}/${user_name}`, this.apiUrl);
    
        return this.http.delete(url.href, { headers: this.headers });
    }

    updateUserRequest(user_name, new_user_name, password, auth_level, email) {
        if (!user_name || !new_user_name || !password || auth_level == null || !email) {
            throw new IncompleteParams();
        }
        const endpoint = "/api/user";
        const url = new URL(`${endpoint}/${user_name}`, this.apiUrl);
        const params = new URLSearchParams();
        params.append('user_name', new_user_name);
        params.append('password', password);
        params.append('auth_level', auth_level);
        params.append('email', email);
    
        return this.http.put(url.href, params, { headers: this.headers });
    }

    getUserRequest() {
        const endpoint = "/api/user";
        const url = new URL(endpoint, this.apiUrl);
    
        return this.http.get(url.href, { headers: this.headers })
            .then(response => {
                if (response.status === 200) {
                    return response.json();
                } else {
                    throw new Error(`Request failed with status ${response.status}`);
                }
            });
    }

    getUserSignupRequests() {
        const endpoint = "/api/user/requests";
        const url = new URL(endpoint, this.apiUrl);
    
        return this.http.get(url.href, { headers: this.headers })
            .then(response => {
                if (response.status === 200) {
                    return response.json();
                } else {
                    throw new Error(`Request failed with status ${response.status}`);
                }
            });
    }

    approveUserRequest(user_name) {
        if (!user_name) {
            throw new IncompleteParams();
        }
        const endpoint = "/api/user/approve";
        const url = new URL(`${endpoint}/${user_name}`, this.apiUrl);
    
        return this.http.post(url.href, null, { headers: this.headers });
    }

    getBlockedUsersRequest() {
        const endpoint = "/api/user/blocked";
        const url = new URL(endpoint, this.apiUrl);
    
        return this.http.get(url.href, { headers: this.headers })
            .then(response => {
                if (response.status === 200) {
                    return response.json();
                } else {
                    throw new Error(`Request failed with status ${response.status}`);
                }
            });
    }

    blockUserRequest(user_name) {
        if (!user_name) {
            throw new IncompleteParams();
        }
        const endpoint = "/api/user/blocked";
        const url = new URL(`${endpoint}/${user_name}`, this.apiUrl);
    
        return this.http.post(url.href, null, { headers: this.headers });
    }

    unblockUserRequest(user_name) {
        if (!user_name) {
            throw new IncompleteParams();
        }
        const endpoint = "/api/user/blocked";
        const url = new URL(`${endpoint}/${user_name}`, this.apiUrl);
    
        return this.http.delete(url.href, { headers: this.headers });
    }

    getDownloadsUserRequest(limit = 1) {
        const endpoint = "/api/user/downloads";
        const url = new URL(`${endpoint}/${limit}`, this.apiUrl);
    
        return this.http.get(url.href, { headers: this.headers })
            .then(response => {
                if (response.status === 200) {
                    return response.json();
                } else {
                    throw new Error(`Request failed with status ${response.status}`);
                }
            });
    }

    getToptenHeaviestUsers() {
        const endpoint = "/api/user/heavy";
        const url = new URL(endpoint, this.apiUrl);
    
        return this.http.get(url.href, { headers: this.headers })
            .then(response => {
                if (response.status === 200) {
                    return response.json();
                } else {
                    throw new Error(`Request failed with status ${response.status}`);
                }
            });
    }

    // Download Functions

    startDownload(server_key = "123456789") {
        const endpoint = "/api/download/start";
        const url = new URL(endpoint, this.apiUrl);
    
        this.headers['key'] = server_key;
    
        return this.http.get(url.href, { headers: this.headers })
            .then(response => {
                if (response.status === 200) {
                    return response.json();
                } else {
                    throw new Error(`Request failed with status ${response.status}`);
                }
            });
    }

    killDownload(server_key = "123456789") {
        const endpoint = "/api/download/kill";
        const url = new URL(endpoint, this.apiUrl);
    
        this.headers['key'] = server_key;
    
        return this.http.get(url.href, { headers: this.headers })
            .then(response => {
                if (response.status === 200) {
                    return response.json();
                } else {
                    throw new Error(`Request failed with status ${response.status}`);
                }
            });
    }

    addDownloadRequest(download_link) {
        if (!download_link) {
            throw new IncompleteParams();
        }
        const endpoint = "/api/download";
        const url = new URL(endpoint, this.apiUrl);
        const params = { link: download_link };
    
        return this.http.post(url.href, JSON.stringify(params), { headers: this.headers })
            .then(response => {
                if (response.status !== 200) {
                    throw new Error("Add download was not successful");
                }
                return response.json();
            });
    }

    removeDownloadRequest(id) {
        if (!id) {
            throw new Error("ID is required to remove download");
        }
        const endpoint = `/api/download/${id}`;
        const url = new URL(endpoint, this.apiUrl);
    
        return this.http.delete(url.href, { headers: this.headers })
            .then(response => {
                if (response.status !== 200) {
                    throw new Error(`Remove download failed with status ${response.status}`);
                }
                return response.json();
            });
    }

    rateDownloadRequest(id, rate) {
        if (!id || !rate) {
            throw new Error("ID and rate are required to rate download");
        }
        const endpoint = `/api/download/${id}`;
        const url = new URL(endpoint, this.apiUrl);
        const params = { rate: rate };
    
        return this.http.post(url.href, JSON.stringify(params), { headers: this.headers })
            .then(response => {
                if (response.status !== 200) {
                    throw new Error(`Rate download failed with status ${response.status}`);
                }
                return response.json();
            });
    }

    getDownloadsRequest(limit) {
        if (limit === undefined || limit === null) {
            throw new Error("Limit is required to get downloads");
        }
        const endpoint = `/api/downloads/${limit}`;
        const url = new URL(endpoint, this.apiUrl);
    
        return this.http.get(url.href, { headers: this.headers })
            .then(response => {
                if (response.status === 200) {
                    return response.json();
                } else {
                    throw new Error(`Get downloads failed with status ${response.status}`);
                }
            });
    }

    getDownload(id) {
        if (!id) {
            throw new Error("ID is required to get download");
        }
        const endpoint = `/api/download/${id}`;
        const url = new URL(endpoint, this.apiUrl);
    
        return this.http.get(url.href, { headers: this.headers })
            .then(response => {
                if (response.status === 200) {
                    return response.json();
                } else {
                    throw new Error(`Get download failed with status ${response.status}`);
                }
            });
    }

    // File Functions

    startCompression(gidList) {
        if (!gidList || gidList.length === 0) {
            throw new Error("GID list is required to start compression");
        }
        const endpoint = "/api/compress";
        const url = new URL(endpoint, this.apiUrl);
        const params = { gid: gidList };
    
        return this.http.post(url.href, JSON.stringify(params), { headers: this.headers })
            .then(response => {
                if (response.status === 200) {
                    return response.json();
                } else {
                    throw new Error(`Start compression failed with status ${response.status}`);
                }
            });
    }

    getCompressionProgress(id) {
        if (!id) {
            throw new Error("ID is required to get compression progress");
        }
        const endpoint = `/api/compression-progress/${id}`;
        const url = new URL(endpoint, this.apiUrl);
    
        return this.http.get(url.href, { headers: this.headers })
            .then(response => {
                if (response.status === 200) {
                    return response.json();
                } else {
                    throw new Error(`Get compression progress failed with status ${response.status}`);
                }
            });
    }

    sendFileFromPath(id) {
        if (!id) {
            throw new Error("ID is required to send file from path");
        }
        const endpoint = "/api/file";
        const url = new URL(endpoint, this.apiUrl);
        url.searchParams.append('gid', id);
    
        return this.http.get(url.href, { headers: this.headers })
            .then(response => {
                if (response.status === 200) {
                    return response.json();
                } else {
                    throw new Error(`Send file from path failed with status ${response.status}`);
                }
            });
    }
    
}

module.exports = Bassa